src/Diagnostics.DataProviders/KustoSDKClient.cs

```diff
@@ -117,7 +129,86 @@ public async Task<DataTable> ExecuteQueryAsync(string query, string cluster, str
+            Exception attemptException = default;
+            var exceptions = new List<Exception>();
+
+            do
+            {
+                try
+                {
+                    if(attempt == _config.MaxRetryCount && _config.MaxRetryCount != 0 && _config.UseBackupClusterForLastRetryAttempt)
+                    {
+                        // Last Retry Attempt
+                        // Switch to backup cluster since useBackupClusterForLastAttempt is set to true.
+                        cluster = GetBackupClusterName(cluster);
+                    }
+
+                    DiagnosticsETWProvider.Instance.LogRetryAttemptMessage(
+                        requestId ?? string.Empty,
+                        source,
+                        $"Starting Attempt : {attempt}, Cluster: {cluster}, Query : {query}"
+                        );
+
+                    invocationStartTime = DateTime.UtcNow;
+                    attemptException = null;
+                    executeQueryTask = ExecuteQueryAsync(query, cluster, database, DataProviderConstants.DefaultTimeoutInSeconds, requestId, operationName);
+                    dtResult = await executeQueryTask;
+                }
+                catch (Exception ex)
+                {
+                    attemptException = ex;
+                    exceptions.Add(ex);
+                }
+                finally
+                {
+                    invocationEndTime = DateTime.UtcNow;
+                    var totalResponseTime = invocationEndTime - invocationStartTime;
+                    string exceptionType = attemptException != null ? attemptException.GetType().ToString() : string.Empty;
+                    string exceptionDetails = attemptException != null ? attemptException.ToString() : string.Empty;
+
+                    DiagnosticsETWProvider.Instance.LogRetryAttemptSummary(
+                            requestId ?? string.Empty,
+                            source,
+                            $"Attempt : {attempt},  IsSuccessful : {attemptException == null}, Cluster: {cluster}, Query : {query}",
+                            Convert.ToInt64(totalResponseTime.TotalMilliseconds),
+                            invocationStartTime.ToString("HH:mm:ss.fff"),
+                            invocationEndTime.ToString("HH:mm:ss.fff"),
+                            exceptionType,
+                            exceptionDetails
+                            );
+
+                    // Logic to check if retry needs to continue
+                    if(totalResponseTime.TotalSeconds > _config.MaxFailureResponseTimeInSecondsForRetry || !IsExceptionRetryable(attemptException))
+                    {
+                        DiagnosticsETWProvider.Instance.LogRetryAttemptMessage(
+                        requestId ?? string.Empty,
+                        source,
+                        $"Not continuing Retries after Attempt : {attempt}, Cluster: {cluster}, Query : {query}"
```

Comment:
>Not continuing Retries [](start = 26, length = 22)

Should we also log totalResponseTime.TotalSeconds ?

Reply:
Its already logged as part of attempt summary event just above.  But i can add it to this event as well

Reply:
Never mind, I see it is logged above...

---
In reply to: [434590128](https://github.com/Azure/Azure-AppServices-Diagnostics/pull/514#discussion_r434590128) [](ancestors = 434590128)